src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -2828,6 +2828,10 @@ void InstanceKlass::release_C_heap_structures(bool release_sub_metadata) {
   jmethodID* jmeths = methods_jmethod_ids_acquire();
   if (jmeths != (jmethodID*)nullptr) {
     release_set_methods_jmethod_ids(nullptr);
+    size_t count = (size_t)jmeths[0];
+    for (size_t i = 1; i <= count; i++) {
+      *((Method**)jmeths[i]) = nullptr;
+    }
```

Comment:
The proposed change looks suspicious: the structure being zeroed is released on the very next line.
So, this change does not seem to fix anything - it only makes a bug (if any) less obvious.
I would rather do the opposite: poison the memory to make sure any attempt to access freed memory will immediately fail and thus reveal the bug in the code.

P.S. JVM TI functions do not access this structure; they use jmethodID cache in ClassLoaderData which is cleared but never released, see https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/classfile/classLoaderData.cpp#L609C14-L622

Reply:
Regarding PS - are you sure?
Eg. when I look at the way the stacktrace is obtained at https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/prims/jvmtiEnvBase.cpp#L1171, it calls to https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/oops/method.cpp#L2196 which in turn calls to https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/oops/instanceKlass.cpp#L2247 and then https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/oops/instanceKlass.cpp#L2397 

This really seems like the the jmethodID array obtained via `methods_jmethod_ids_acquire()` is the source of truth, at least in this case.

Regarding zeroing the released structure - IIUC, by doing something like `jmethodID mid = jmeths[idx]` the `mid` value (which is in fact a pointer to `Method`) will remain valid even after the holding array has been deallocated.

I have also did a small experiment, just to be sure I am not making a complete fool of myself with following code
```
mid = jni->GetStaticMethodID(clz, "run", "()V");

if (mid != nullptr) {
  printf("jmethodID points to null: %s\n", *((void**)mid) == nullptr ? "true" : "false");
  jmethodID mid1 = mid;
  *((void**)mid1) = nullptr;
  printf("jmethodID points to null: %s\n", *((void**)mid) == nullptr ? "true" : "false");
}
```
yielding
```
jmethodID points to null: false
jmethodID points to null: true
```

I am running some extra tests now and will switch the PR to draft until I have some results

Reply:
A `jmethodID` is just a pointer to a `Method*` - essentially the address of a `Method*` array entry. What you are proposing is that as part of the unloading of the class when we tear down these structures, we null that array entry so that a null check could be performed on an existing `jmethodID`. But "shortly" after we null the entry the entire array can be released and the `jmethodID` is now a dangling pointer.

Reply:
@jbachorik Sorry for my confusing statement. Of course, this structure _is_ used by JVM TI, but what I meant is that it's just a jmethodID cache; it is not used to _resolve_ jmethodIDs. The actual jmethodID slots returned to user are allocated in `JNIMethodBlock` structures belonging to `ClassLoaderData` and they are never released. See `Method::make_jmethod_id`.

Reply:
I am pretty sure this is true
```
jmethodID mid1 = <get jmethodID>
jmethodID* arr = (jmethodID*)malloc(1 * sizeof(jmethodID));
arr[0] = mid; // arr[0] is a copy of mid1
jmethodID mid2 = arr[0]; // mid2 is a copy of arr[0]
memset(arr, 3, 1 * sizeof(jmethodID)); // the contents of array are overwritten by 3
free(arr); // and now the array memory is released

// yet, when we do this
jvmt->GetMethodName(mid2, &mname, &msig, NULL);
// the returned mname and msig are correct and no JVMTI error is reported as I would expect if I overwrite the jmethodID by garbage
```

In addition, I am not really nulling the array slot - it is rather nulling the pointer to `Method` which that array slot represents.

@apangin 
I am not arguing that the `jmethodID`s are or are not released - my problem is that even though they are not released the `Method` objects they are pointing to can and will be, causing potential access to freed memory.

Considering the previous code snippet, there is also a distinction between doing
```
arr[0] = nullptr; // does not affect to which Method the mid2 points to
```
and
```
*((void**)arr[0]) = nullptr; // this will change the Method ptr for all copies of this particular jmethodID
```

I have verified this experimentally.

Reply:
@coleenp 
I think the jmethodID is copied - see https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/oops/instanceKlass.cpp#L2397

We are not passing references to jmethodID but rather values. The stacktrace frame will contains a copy of jmethodID which will point to a deallocated Method instance once that holding class is unloaded.

Anyway, we do see inexplicable crashes related to the JVMTI stacktraces (this is just inferred - we see the crashes to cease immediately after we disable the functionality using the JVMTI stacktraces) and this is the only place I could find which looked at least a bit suspicious - all other places are proactively 'destroying' the jmethodID values such that they can be checked for validity before dereferencing the Method.

Thanks everyone for looking into this - I will try to have a reproducer (even though it might not be a simple jtreg test) and try to validate the assumption that the crash we are observing will be fixed. Until then I will keep in draft state.

Reply:
@jbachorik I understand you are zeroing `Method*` pointers, not elements of `jmethodID[]` array. But that's exactly what the [code](https://github.com/openjdk/jdk/blob/87b08b6e0192d88025c2275c7dd2c4bdecda58e8/src/hotspot/share/classfile/classLoaderData.cpp#L609-L622) I linked above does. Can you please explain a scenario where cleanup in `Method::clear_jmethod_ids` does not work, but your fix helps? If there is such a case, I believe the code should be refactored anyway to leave only one place for zeroing jmethodIDs.

Also, what makes me believe the actual bug is in `libjavaProfiler` rather than in HotSpot, is the fact that according to hs_err.log, `jvmti_GetLineNumberTable` is called with `jmethodID == 0xc` which cannot be a pointer at all, even a stale one.

Reply:
@apangin Yes, I get your point. Sorry for being a bit stubborn. I have re-sifted the code and my change is indeed a no-op.
However, thanks to this discussion I actually find 2 code paths leading to deallocation of method (metadata) without doing the jmethodid cleanup.
```
ServiceThread -> ClassLoaderDataGraph::safepoint_and_clean_metaspaces() -> VM_CleanClassLoaderDataMetaspaces -> ClassLoaderDataGraph::walk_metadata_and_clean_metaspaces() -> ClassLoaderDataGraph::clean_deallocate_lists() -> ClassLoaderData::free_deallocate_list() -> MetadataFactory::free_metadata() -> InstanceKlass::deallocate_contents() 
```
```
JvmtiEnv::RedefineClasses/RetransformClasses -> VM_RedefineClasses -> ClassLoaderDataGraph::clean_deallocate_lists() -> ClassLoaderData::free_deallocate_list() -> MetadataFactory::free_metadata() -> InstanceKlass::deallocate_contents() -> InstanceKlass::release_C_heap_structures() 
```

Regarding the hs_err.log - you are right, I missed the slightly different mode of crash. I will try and dig out a crash log showing the problem when dereferencing a method.